### ISSUE TITLE
TaskWorkers running in parallel conflict on submissions

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#5719]
When two task workers run in paralled on same DB we found (at lest) one case where apparently the select ad lock function in the DB did not work, resulting in both machines working on same task at basically same time.
See attached TW logs.
When submitting somehow the duplicate check avoided two actual condor submissions, but the end result was that the dagBoostrap ClusterId was recored as None in the DB.

The DB content (from crabserver UI) for the task is:
`

name | value
-- | --
tm_taskname | 180821_092838:jdegens_crab_SingleMuon_Run2017C_31Mar2018v1_13TeV_MINIAOD
tm_activity | None
panda_jobset_id | None
tm_task_status | SUBMITTED
tm_task_command | SUBMIT
tm_start_time | 2018-08-21 09:28:39.270621
tm_start_injection | None
tm_end_injection | 2018-08-21 09:30:18.642910
tm_task_failure | None
tm_job_sw | CMSSW_9_4_9
tm_job_arch | slc7_amd64_gcc630
tm_input_dataset | /SingleMuon/Run2017C-31Mar2018-v1/MINIAOD
tm_nonvalid_input_dataset | F
tm_use_parent | 0
tm_site_whitelist | []
tm_site_blacklist | ['T3_FR_IPNL',  'T2_RU_IHEP', 'T3_US_UCR', 'T1_RU_JINR', 'T2_EE_Estonia', 'T2_RU_SINP',  'T2_RU_INR', 'T2_RU_PNPI', 'T3_US_UMiss', 'T2_RU_JINR', 'T2_TW_NCHC',  'T2_RU_ITEP', 'T2_BR_SPRACE', 'T2_US_Vanderbilt']
tm_split_algo | Automatic
tm_split_args | {'splitOnRun': False, 'minutes_per_job': 480, 'halt_job_on_file_boundaries': False, 'runs': [], 'lumis': []}
tm_user_sandbox | 91bec16ebe7f4257b41bbc8b7888521b553609c0fc5438c85f22c7d8a99fbe9d.tar.gz
tm_cache_url | https://cmsweb.cern.ch/crabcache
tm_username | jdegens
tm_user_dn | /C=DE/O=GermanGrid/OU=RWTH/CN=Jordy Degens
tm_user_vo | cms
tm_user_role | None
tm_user_group | dcms
tm_publish_name | crab_SingleMuon_Run2017C_31Mar2018v1_13TeV_MINIAOD-00000000000000000000000000000000
tm_publish_groupname | F
tm_asyncdest | T2_DE_RWTH
tm_dbs_url | https://cmsweb.cern.ch/dbs/prod/global/DBSReader
tm_publish_dbs_url | https://cmsweb.cern.ch/dbs/prod/phys03/DBSWriter
tm_publication | F
tm_outfiles | ['kappa_SingleMuon_Run2017C_31Mar2018v1_13TeV_MINIAOD.root']
tm_tfile_outfiles | []
tm_edm_outfiles | []
tm_job_type | Analysis
tm_generator | pythia
tm_events_per_lumi | None
tm_arguments | {}
panda_resubmitted_jobs | []
tm_save_logs | T
tw_name | crab-prod-tw01
tm_user_infiles | []
tm_maxjobruntime | 1250
tm_numcores | 1
tm_maxmemory | 2500
tm_priority | 10
tm_output_dataset | None
tm_task_warnings | []
tm_user_webdir | http://vocms0195.cern.ch/mon/cms1708/180821_092838:jdegens_crab_SingleMuon_Run2017C_31Mar2018v1_13TeV_MINIAOD
tm_scriptexe | None
tm_scriptargs | []
tm_extrajdl | []
tm_asourl | https://cmsweb.cern.ch/crabserver/prod
tm_collector | cmsgwms-collector-global.cern.ch:9620,cmssrv221.fnal.gov:9620
tm_schedd | crab3@vocms0195.cern.ch
tm_dry_run | F
tm_user_files | []
tm_transfer_outputs | T
tm_output_lfn | /store/user/jdegens/higgs-kit/skimming/2018_08_21_datatry2
tm_ignore_locality | F
tm_fail_limit | None
tm_one_event_mode | F
tm_secondary_input_dataset | None
tm_primary_dataset | SingleMuon
tm_last_publication | None
tm_debug_files | 4ebd7a45a1e979d08a67a8a605a144fa6c062471e057158bbbaec53c827d7eda.tar.gz
clusterid | None
tm_asodb | filetransfers
tm_ignore_global_blacklist | F
tm_submitter_ip_addr | 134.61.15.83
tm_totalunits | 0.0

``

```

**Comment by belforte** [comment]
More details.
the TW name recorded in taskDB is crab-prod-tw01 which corresponds to vocms052, the other TW running at the moment was crab--prod-tw02.
The log files for this task from the two TW's are uploaded here:

[tw01.log](https://github.com/dmwm/CRABServer/files/2310600/tw01.log)
[tw02.log](https://github.com/dmwm/CRABServer/files/2310601/tw02.log)

**Comment by belforte** [comment]
There are two problems here:
[ ] the same task was picked by two task workers (which is very bad)
[ ] when the duplicate check detected that task was already submitted, it overwrote the ClusterID in the database with None (which is sort of a normal bug)

I should be able to fix the latter, but the first one may be beyond scope and need, at the very least it would require an extensive testing program

**Comment by belforte** [comment]
More about the first problem.
Oddily but somehow consistetly, master log from both machines shows that that they think they own the task:

```
belforte@vocms052/jdegens> grep 180821_092838:jdegens_crab_SingleMuon_Run2017C_31Mar2018v1_13TeV_MINIAOD ../../processes/proc.c3id_master.pid_*2018-08-21|tr  ',' '\n'|grep tw_name
 u'tw_name': u'crab-prod-tw01'
 u'tw_name': u'crab-prod-tw01'
belforte@vocms052/jdegens> 

belforte@crab-prod-tw02/jdegens> grep 180821_092838:jdegens_crab_SingleMuon_Run2017C_31Mar2018v1_13TeV_MINIAOD ../../processes/proc.c3id_master.pid_*2018-08-21|tr  ',' '\n'|grep tw_name
 u'tw_name': u'crab-prod-tw02'
 u'tw_name': u'crab-prod-tw02'
belforte@crab-prod-tw02/jdegens> 

```

**Comment by belforte** [comment]
for the record, the full info on each master process is
[master01.log](https://github.com/dmwm/CRABServer/files/2311011/master01.log)
[master02.log](https://github.com/dmwm/CRABServer/files/2311012/master02.log)


and by prettyfying and diffing I get
[master01-pretty.log](https://github.com/dmwm/CRABServer/files/2311014/master01-pretty.log)
[master02-pretty.log](https://github.com/dmwm/CRABServer/files/2311015/master02-pretty.log)
and
```
LapSB:~$ cat master01.log |tr ',' '\n' > master01-pretty.log
LapSB:~$ cat master02.log |tr ',' '\n' > master02-pretty.log
LapSB:~$ diff master01-pretty.log master02-pretty.log 
79c79
<  u'tw_name': u'crab-prod-tw01'
---
>  u'tw_name': u'crab-prod-tw02'
171c171
<  u'tw_name': u'crab-prod-tw01'
---
>  u'tw_name': u'crab-prod-tw02'
185,186c185,186
< 2018-08-21 11:28:52
< 746:INFO:Worker:Injecting work 9: 180821_092838:jdegens_crab_SingleMuon_Run2017C_31Mar2018v1_13TeV_MINIAOD
---
> 2018-08-21 11:28:51
> 705:INFO:Worker:Injecting work 7: 180821_092838:jdegens_crab_SingleMuon_Run2017C_31Mar2018v1_13TeV_MINIAOD
LapSB:~$ 

```

**Comment by belforte** [comment]
@mmascher  do you have any idea here ?
if I  were to take this at face value, I'd say that Oracle UPDATE is not atomic
 
https://github.com/dmwm/CRABServer/blob/3.3.1809.rc5/src/python/Databases/TaskDB/Oracle/Task/Task.py#L141

But since we can't believe such a bug in Oracle... I am very puzzled and inclined to think that
we will never understand this unless we can reliably reproduce it, i.e. a lot of effort which I
do not see anybody around with the time and competence to do.
So I suspect it will stay like this forever.

My only goal at this point is to make sure that I collect the available evidence from the machines before logs are wiped out. And move on.

**Comment by belforte** [comment]
more  bits:

from master worker on vocms052 = crab-prod-tw01, the task was retrieved from the queue at
```
2018-08-21 11:28:51,931:INFO:MasterWorker:Retrieved a total of 2 works
2018-08-21 11:28:51,931:DEBUG:MasterWorker:Retrieved the following works: 
```
the other task retrieved at same time is
180821_092814:jdegens_crab_SingleMuon_Run2017B_31Mar2018v1_13TeV_MINIAOD

given the source code [*] it means _lockWork had been called right before, but there's no time stamp in log, anyhow after the last call to _getWork which was 23.8 sec earlier:
```
2018-08-21 11:28:13,448:INFO:MasterWorker:Retrieved a total of 3 works
```
while on crab-prod-tw02 the probem task `180821_092838:jdegens_crab_SingleMuon_Run2017C_31Mar2018v1_13TeV_MINIAOD` was retrieved at ~same time 
```
2018-08-21 11:28:51,173:INFO:MasterWorker:Retrieved a total of 2 works
2018-08-21 11:28:51,173:DEBUG:MasterWorker:Retrieved the following works: 
```
again together with task `180821_092814:jdegens_crab_SingleMuon_Run2017B_31Mar2018v1_13TeV_MINIAOD`
and again previous iteration was  at about same time as vocms052:
```
2018-08-21 11:28:13,014:INFO:MasterWorker:Retrieved a total of 3 works
```

Was the second task similarly affected ?



[*]
```

        while(not self.STOP):
            limit = self.slaves.queueableTasks()
            if not self._lockWork(limit=limit, getstatus='NEW', setstatus='HOLDING'):
                time.sleep(self.config.TaskWorker.polling)
                continue

            pendingwork = self._getWork(limit=limit, getstatus='HOLDING')

            if len(pendingwork) > 0:
                self.logger.info("Retrieved a total of %d works", len(pendingwork))
                self.logger.debug("Retrieved the following works: \n%s", str(pendingwork))
```

**Comment by belforte** [comment]
about the second task retrieved at same time by both TW's:
* on crab-prod-tw02:
```
Last successful submission: ClusterId 26190441 submitted to schedd crab3@vocms0137.cern.ch

2018-08-21 11:29:18,634:INFO:Handler:Finished <TaskWorker.Actions.DagmanSubmitter.DagmanSubmitter object at 0x7f18b4a51590> on 180821_092814:jdegens_crab_SingleMuon_Run2017B_31Mar2018v1_13TeV_MINIAOD in 6 seconds
```
* on crab-prod-tw01 = vocms052:
```
Last successful submission: ClusterId 26917824 submitted to schedd crab3@vocms0155.cern.ch

2018-08-21 11:29:14,145:INFO:Handler:Finished <TaskWorker.Actions.DagmanSubmitter.DagmanSubmitter object at 0x7f211c258190> on 180821_092814:jdegens_crab_SingleMuon_Run2017B_31Mar2018v1_13TeV_MINIAOD in 5 seconds
```

this time the two TW's picked two different schedd's and the duplicateCheck could not help, task was submitted twice !
tw01 put is name in the DB 
https://cmsweb.cern.ch/crabserver/ui/task/180821_092814%3Ajdegens_crab_SingleMuon_Run2017B_31Mar2018v1_13TeV_MINIAOD
but the stored ClusterId in the DB is 26190441 i.e. the job sbmitted by tw02 and indeed also the recorded schedd is the one from the submission by tw02: vocms0137.cern.ch.

Clearly the tw_name field in DB is only written at _lockWork time and not changed anymore.

Bloody mess.

**Comment by belforte** [comment]
About this new task submitted by two schedd's, dashboard apparently only has record of jobs from vocms0137. Where task status is all fine:
```
belforte@vocms0137/~> condor_q 26190441 -af crab_reqname
180821_092814:jdegens_crab_SingleMuon_Run2017B_31Mar2018v1_13TeV_MINIAOD
belforte@vocms0137/~> condor_q 26190441


-- Schedd: crab3@vocms0137.cern.ch : <188.184.90.114:4080?... @ 08/22/18 22:56:12
 ID          OWNER            SUBMITTED     RUN_TIME ST PRI SIZE CMD
26190441.0   cms1708         8/21 11:29   0+00:30:20 C  10   0.1 dag_bootstrap_startup.sh RunJobs.da

1 jobs; 1 completed, 0 removed, 0 idle, 0 running, 0 held, 0 suspended
belforte@vocms0137/~> 
```

Yet the task has been progressing nicely also on vocms0155
```
belforte@vocms0155/cluster26917824.proc0.subproc0> condor_q 26917824 -af crab_reqname
180821_092814:jdegens_crab_SingleMuon_Run2017B_31Mar2018v1_13TeV_MINIAOD
belforte@vocms0155/cluster26917824.proc0.subproc0> condor_q 26917824                 


-- Schedd: crab3@vocms0155.cern.ch : <188.184.95.12:4080?... @ 08/22/18 22:55:06
 ID          OWNER            SUBMITTED     RUN_TIME ST PRI SIZE CMD
26917824.0   cms1708         8/21 11:29   0+00:35:24 C  10   0.1 dag_bootstrap_startup.sh RunJobs.da

1 jobs; 1 completed, 0 removed, 0 idle, 0 running, 0 held, 0 suspended
belforte@vocms0155/cluster26917824.proc0.subproc0> 
```

So we have no other "accidental" protection but the duplicate check. Once two TW's pick the same task, it will be run in parallel in two schedd's.

**Comment by belforte** [comment]
now look in REST logs for the overlapping _lockWork calls.
There are four backends for production crabserver at the moment: vocms0136 vocms0161 vocms0163 vocms0165 (from https://cms-http-group.web.cern.ch/cms-http-group/activity.html)
I look for the `UPDATE tasks SET tw_name = :tw_name, tm_task_status = :set_status` (from https://github.com/dmwm/CRABServer/blob/3.3.1809.rc5/src/python/Databases/TaskDB/Oracle/Task/Task.py#L141 ) at 11:28 of August 21
`grep ":11:28"  vocms0161/crabserver/crabserver-20180821.log|grep "UPDATE tasks SET tw_name"`


* vocms0136 : nothing
* vocms0161 : one call
[C1]
```
[21/Aug/2018:11:28:44]  RESTSQL:TZjzfyBOMEhU prepare [UPDATE tasks SET tw_name = :tw_name, tm_task_status = :set_status                          WHERE tm_taskname IN (SELECT tm_taskname FROM (SELECT tm_taskname, rownum as counter                          FROM tasks WHERE tm_task_status = :get_status ORDER BY tm_start_time)                          WHERE counter <= :limit)]
```
* vocms0163 : two calls
[C2]
```
[21/Aug/2018:11:28:06]  RESTSQL:HyAXtlnvEDQw prepare [UPDATE tasks SET tw_name = :tw_name, tm_task_status = :set_status                          WHERE tm_taskname IN (SELECT tm_taskname FROM (SELECT tm_taskname, rownum as counter                          FROM tasks WHERE tm_task_status = :get_status ORDER BY tm_start_time)                          WHERE counter <= :limit)]
```
[C3]
```
[21/Aug/2018:11:28:45]  RESTSQL:AXnGCRidzFOP prepare [UPDATE tasks SET tw_name = :tw_name, tm_task_status = :set_status                          WHERE tm_taskname IN (SELECT tm_taskname FROM (SELECT tm_taskname, rownum as counter                          FROM tasks WHERE tm_task_status = :get_status ORDER BY tm_start_time)                          WHERE counter <= :limit)]
```

* vocms0165 :  one call
[C4]
```
[21/Aug/2018:11:28:06]  RESTSQL:oRJxuzbXZTcg prepare [UPDATE tasks SET tw_name = :tw_name, tm_task_status = :set_status                          WHERE tm_taskname IN (SELECT tm_taskname FROM (SELECT tm_taskname, rownum as counter                          FROM tasks WHERE tm_task_status = :get_status ORDER BY tm_start_time)                          WHERE counter <= :limit)]
```

I guess that [C1] and [C3] are the ones related to the work retrieval at 11:28:51
Now checking for more details in logs.

**Comment by belforte** [comment]
one crazy idea... is is possible that the same task name appeared twice in the database ? I.e. the user was running two identical submit at same time ? That may lead to above scenario even if DB calls are perfectly atomic and safe.

**Comment by belforte** [comment]
that was crazy indeed. No duplicates:
```
belforte@stefanovm2/CRABServer> curl -s --key /tmp/x509up_u8516 --cert /tmp/x509up_u8516 -k  -X GET 'https://cmsweb.cern.ch/crabserver/prod/task?subresource=taskbystatus&username=jdegens&taskstatus=SUBMITTED'|grep 180821|sort|uniq -c
      1 ,["SUBMITTED", "180821_084615:jdegens_crab_SingleElectron_Run2017B_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_084637:jdegens_crab_SingleElectron_Run2017C_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_084658:jdegens_crab_SingleElectron_Run2017D_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_084720:jdegens_crab_SingleElectron_Run2017E_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_084741:jdegens_crab_SingleElectron_Run2017F_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_084803:jdegens_crab_SingleMuon_Run2017B_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_084826:jdegens_crab_SingleMuon_Run2017C_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_084849:jdegens_crab_SingleMuon_Run2017D_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_084912:jdegens_crab_SingleMuon_Run2017E_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_084936:jdegens_crab_SingleMuon_Run2017F_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_092622:jdegens_crab_SingleElectron_Run2017B_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_092643:jdegens_crab_SingleElectron_Run2017C_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_092706:jdegens_crab_SingleElectron_Run2017D_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_092729:jdegens_crab_SingleElectron_Run2017E_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_092751:jdegens_crab_SingleElectron_Run2017F_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_092814:jdegens_crab_SingleMuon_Run2017B_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_092838:jdegens_crab_SingleMuon_Run2017C_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_092902:jdegens_crab_SingleMuon_Run2017D_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_092926:jdegens_crab_SingleMuon_Run2017E_31Mar2018v1_13TeV_MINIAOD"]
      1 ,["SUBMITTED", "180821_092952:jdegens_crab_SingleMuon_Run2017F_31Mar2018v1_13TeV_MINIAOD"]
belforte@stefanovm2/CRABServer> 

```

**Comment by belforte** [comment]
All I can do at this point is to downgrade priority to low, and assume that we will work on this someday only if running multiple TW's in parallel becomes a pressing need.

**Comment by belforte** [comment]
After investigation by Marco and Todor, this is indeed a bug in Oracle which is triggered by the rather complex SELECT that we do to order the tasks by date.
We decided to simplyfy the query. Followed up in the new issue #5761
